src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -0,0 +1,199 @@
+ private:
+  debug_only(static intx _locking_thread_id;)
+  static Semaphore _lock;
+ public:
+  AsyncLogLocker() {
+    _lock.wait();
+    debug_only(_locking_thread_id = os::current_thread_id());
+  }
+
+  ~AsyncLogLocker() {
+    debug_only(_locking_thread_id = -1);
+    _lock.signal();
+  }
+};
+
+Semaphore AsyncLogLocker::_lock(1);
+debug_only(intx AsyncLogLocker::_locking_thread_id = -1;)
+
+void AsyncLogWriter::enqueue_locked(const AsyncLogMessage& msg) {
+  if (_buffer.size() >= _buffer_max_size)  {
+    const AsyncLogMessage* h = _buffer.front();
+    assert(h != NULL, "sanity check");
+
+    if (h->message() != nullptr) {
+      bool p_created;
+      uint32_t* counter = _stats.add_if_absent(h->output(), 0, &p_created);
+      *counter = *counter + 1;
+    }
+    // drop the enqueueing message.
+    return;
+  }
+
+  assert(_buffer.size() < _buffer_max_size, "_buffer is over-sized.");
+  _buffer.push_back(msg);
+  _sem.signal();
+}
+
+void AsyncLogWriter::enqueue(LogFileOutput& output, const LogDecorations& decorations, const char* msg) {
+  AsyncLogMessage m(output, decorations, os::strdup(msg));
+
+  { // critical area
+    AsyncLogLocker lock;
+    enqueue_locked(m);
+  }
+}
+
+// LogMessageBuffer consists of a multiple-part/multiple-line messsages.
+// the mutex here gurantees its integrity.
+void AsyncLogWriter::enqueue(LogFileOutput& output, LogMessageBuffer::Iterator msg_iterator) {
+  AsyncLogLocker lock;
+
+  for (; !msg_iterator.is_at_end(); msg_iterator++) {
+    AsyncLogMessage m(output, msg_iterator.decorations(), os::strdup(msg_iterator.message()));
+    enqueue_locked(m);
+  }
+}
+
+AsyncLogWriter::AsyncLogWriter()
+  : _state(ThreadState::NotReady),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    _state = ThreadState::Initialized;
+  }
+
+  log_info(logging)("The maximum entries of AsyncLogBuffer: " SIZE_FORMAT ", estimated memory use: " SIZE_FORMAT " bytes",
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+class AsyncLogMapIterator {
+  AsyncLogBuffer& _logs;
+
+ public:
+  AsyncLogMapIterator(AsyncLogBuffer& logs) :_logs(logs) {}
+  bool do_entry(LogFileOutput* output, uint32_t* counter) {
+    using none = LogTagSetMapping<LogTag::__NO_TAG>;
+
+    if (*counter > 0) {
+      LogDecorations decorations(LogLevel::Warning, none::tagset(), output->decorators());
+      stringStream ss;
+      ss.print(UINT32_FORMAT_W(6) " messages dropped due to async logging", *counter);
+      AsyncLogMessage msg(*output, decorations, ss.as_string(true /*c_heap*/));
+      _logs.push_back(msg);
+      *counter = 0;
+    }
+
+    return true;
+  }
+};
+
+void AsyncLogWriter::perform_IO() {
+  // use kind of copy-and-swap idiom here.
+  // Empty 'logs' 'swaps' the content with _buffer.
+  // Along with logs destruction, all procceeded messages are deleted.
+  //
+  // the atomic operation 'move' is done in O(1). All I/O jobs are done without lock.
```

Comment:
` _buffer.pop_all(&logs)` is placed in 'critical region' on purpose. 
It's because logsites may be enqueuing concurrently. that's why I need to use a lock to protect _buffer. 
I see 'atomic' is inappropriate in the context. I will remove it.